core_impl/src/main/java/com/google/instrumentation/trace/SpanImpl.java

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.trace;
+
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.instrumentation.common.Clock;
+import java.util.Collections;
+import java.util.EnumSet;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.ThreadSafe;
+
+/** Implementation for the {@link Span} class. */
+@ThreadSafe
+final class SpanImpl extends Span {
+  private static final Logger logger = Logger.getLogger(Tracer.class.getName());
+
+  private final SpanId parentSpanId;
+  private final StartEndHandler startEndHandler;
+  private final String name;
+  // All the following variables are initialized iff the Span has RECORD_EVENTS option.
+  private final Clock clock;
+  private TimestampConverter timestampConverter;
```

Comment:
Added comments. Also timestamp converter is used for multiple reasons:
* Better granularity than ms for tracing events.
* Ensure tracing events ordering between multiple spans from the same trace. We use the same converter for all the Spans that are part of the same trace to ensure for example that parent start time < child start time.